src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp

```diff
@@ -395,10 +404,11 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::WriteAttribute(const Dat
     //
     // - Internal writes may need to be able to decide if to mark things dirty or not (see AAI as well)
     // - Changes-ommited paths should not be marked dirty (ember is not aware of that flag)
-    // - This likely maps to `MatterReportingAttributeChangeCallback` HOWEVER current ember write functions
-    //   will selectively call that one depending on old attribute state (i.e. calling every time is a
-    //   change in behavior)
-    CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+    if (attributeChanged == AttributeChanged::kValueChanged)
+    {
+        emberAfIncreaseClusterDataVersion(request.path);
+        CurrentContext().dataModelChangeListener->MarkDirty(request.path);
```

Comment:
Why do we want these changes?  This still requires the ember code to support dirty-marking itself, but now we are supporting it here too, which just adds to the codesize for no good reason.  This part does not make sense to me.

Reply:
Existing ember code at write time is totally unaware of cluster versioning as far as I could tell - it lets cluster version being increased as a sideffect of change callbacks.

This is the bit that I find somewhat hard to follow on the intent: we advise clusters to call `MatterAttributeChangedCallback` which marks dirty and increases the cluster version. However cluster version seems to be a property of clusters/data-model.

Will followup over chat on this since the existing split (on why version is in emberAf ... probably because of convenience of data structure) makes things hard.

Reply:
>  it lets cluster version being increased as a sideffect of change callbacks.

Right, cluster version increases at the place where the data mutates, whether it's due to a write or due to AttributeName::Set, or whatever.

> we advise clusters to call MatterAttributeChangedCallback which marks dirty and increases the cluster version.

We advise _whoever controls changes to the data_ to do that when the data changes.  If the cluster controls those changes, it should do that.

> cluster version seems to be a property of clusters/data-model.

What it's a property of and who is responsible for updating it are different questions.

Reply:
Updated with code changes after #35455